src/webgpu/api/validation/capability_checks/limits/limit_utils.ts

```diff
@@ -0,0 +1,292 @@
+    const limit = this.limit;
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    assert(!!adapter);
+
+    const defaultLimit = (kLimitInfo as Record<string, { default: number }>)[limit].default;
+    const maximumLimit = adapter.limits[limit] as number;
+    assert(!Number.isNaN(defaultLimit));
+    assert(!Number.isNaN(maximumLimit));
+
+    const requestedLimit = getLimitValue(defaultLimit, maximumLimit, limitValueTest);
+
+    const requiredLimits: Record<string, number> = {};
+    requiredLimits[limit] = requestedLimit;
+
+    const device = await this.requestDeviceWithLimits(limitValueTest, adapter, requiredLimits);
+    const actualLimit = (device ? device.limits[limit] : 0) as number;
+
+    switch (limitValueTest) {
+      case 'atDefault':
+        this.expect(!!device);
+        this.expect(actualLimit === defaultLimit);
+        break;
+      case 'underDefault':
+        this.expect(!!device);
+        this.expect(actualLimit === defaultLimit);
+        break;
+      case 'atMaximum':
+        this.expect(!!device);
+        this.expect(actualLimit === maximumLimit);
+        break;
+      case 'overMaximum':
+        this.expect(!device);
+        break;
+    }
+
+    return device ? { device, defaultLimit, maximumLimit, requestedLimit, actualLimit } : undefined;
+  }
+
+  /**
+   * Creates a device with the requested limits.
+   * If the limit of over the maximum we expect an exception
+   * If the device is created then we call a test function, checking
+   * that the function does not leak any GPU errors.
+   */
+  async testDeviceWithRequestedLimits(
+    limitTest: keyof typeof LimitValueTest,
+    testValueName: keyof typeof TestValue,
+    fn: (inputs: LimitsTestFnInputs) => void | Promise<void>
+  ) {
+    const deviceAndLimits = await this._getDeviceWithRequestedLimit(limitTest);
+    // If we request over the limit requestDevice will throw
+    if (!deviceAndLimits) {
+      return;
+    }
+
+    const { device, actualLimit } = deviceAndLimits;
+    this._device = device;
+    const testValue = getTestValue(actualLimit, testValueName);
+    const shouldError = testValueName === 'overLimit';
+
+    device.pushErrorScope('internal');
+    device.pushErrorScope('out-of-memory');
+    device.pushErrorScope('validation');
+
+    const result = fn({ ...deviceAndLimits, testValueName, testValue, shouldError });
+    if (result instanceof Promise) {
+      await result;
+    }
+
+    const validationError = await device.popErrorScope();
+    const outOfMemoryError = await device.popErrorScope();
+    const internalError = await device.popErrorScope();
+
+    this.expect(!validationError, validationError?.message || '');
+    this.expect(!outOfMemoryError, outOfMemoryError?.message || '');
+    this.expect(!internalError, internalError?.message || '');
+
+    device.destroy();
+    this._device = undefined;
+  }
+
+  /**
+   * Calls a function that expects a GPU error if shouldError is true
+   */
```

Comment:
Let's add a MAINTENANCE_TODO to remove this duplicated code with GPUTest if possible (even if we don't know how we'll do it yet - https://github.com/gpuweb/cts/pull/2184#issuecomment-1420055721)

Reply:
done